chore: create docker image and compose for dev using the watch feature

pbrissaud opened the pull request:
This commit create new image and compose file for dev mode. It use new compose feature `watch` released in version **2.22.0 (september 2023)** to automatically reload container when changes are detected

Video demo (password: tamagokill)
https://vimeo.com/952926923?share=copy

There is just a small delay for the API after the change; the time it takes to rebuild.

Don't forget `--watch` flag in docker compose command to enable this.

Example : `docker compose -f compose.dev.yaml up --build --watch`

Resources : 
- Cargo watch : https://crates.io/crates/cargo-watch
- Docker compose watch : https://docs.docker.com/compose/file-watch/

pbrissaud committed:
chore: create docker image and compose for dev using the watch feature

This commit create new image and compose file for dev mode. It use new compose feature `watch` released in version 2.22.0 (september 2023) to automatically reload container when changes are detected

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

lilymonad commented on compose.yaml:
```diff
@@ -6,23 +6,13 @@ services:
-      - target: 443
-        published: ${HTTP3_PORT:-443}
-        protocol: udp
+      - 8080:8080
```
Revert this change, it is totaly OK to have it since the API could be exposed to outside.
If you want to modify default port forwarding just use environment variables on your local configuration like so:

```
HTTP_PORT=54321 docker-compose up
```
(or any other values of HTTP/HTTPS/HTTP3 ports you want)


It would also be nice to have it as an environment variable for the `web` container.

pbrissaud replied:
The API doesn't listen to 443 and 80 so the declaration of these ports are useless  https://github.com/DevGirl-Team/tamagokill/blob/665d9616757c4bce55fe4bbbd7237de688234830/api/src/main.rs#L22

Ok for the local port configuration as env 👍 Will change the code

lilymonad replied:
Agreed on the TLS/HTTP3 ports **but** I think it's not that problematic to keep them for now, since it makes sense for the future.

lilymonad commented on web/package.json:
```diff
@@ -36,5 +36,10 @@
 	"dependencies": {
 		"dotenv": "^16.4.5"
-	}
+	},
+	"overrides": { 
+    "vite": {
+      "rollup": "npm:@rollup/wasm-node"
+    }
+  }
```
I don't understand why re-installing vite.
Didn't we agreed on using svelte @Lauwed ?

pbrissaud replied:
Had a problem when installing the web dependencies on my machine (MacOS) . Found a similar issue and workaround here: https://github.com/vitejs/vite/issues/15167#issuecomment-1833274104

Lauwed replied:
I didn't take the time to read anything right now, but I know that SvelteKit is build on Vite.

pbrissaud replied:
I can delete this change for this PR but I think other contributor on Mac could have this issue

ferdodo replied:
@pbrissaud can you still reproduce the problem with an up to date version of sveltekit ?

pbrissaud replied:
@ferdodo Nope, I can create a new svelte-kit project on my Mac without any issues. 

And now, there are no more errors on this project when I delete the "overrides" part... I will delete it!

ferdodo replied:
@pbrissaud  perfect ! 🎉

lilymonad commented on compose.dev.yaml:
```diff
@@ -0,0 +1,47 @@
+    depends_on:
+      - database
+    ports:
+      - 8080:8080
```
Same comment as for the compose.yam file, just use the same port forwarding layout, and if you want specific ports just use environment variables

lilymonad commented on compose.dev.yaml:
```diff
@@ -0,0 +1,47 @@
+        - action: rebuild
+          path: ./web/package.json
+    ports: 
+      - 3000:5173
```
use environment variables

lilymonad commented on api/Dockerfile.dev:
```diff
@@ -0,0 +1,13 @@
+
+WORKDIR /app
+
+COPY . .
```
Prefer binding at compose.dev.yaml level instead of copying the workspace in the development image.

Also why `cargo install --path .` since it's a ""devcontainer"" ?

pbrissaud replied:
> Prefer binding at compose.dev.yaml level instead of copying the workspace in the development image.

The watch feature in Docker Compose is designed to avoid volume binding due to its performance issues (source: [Docker Blog](https://www.docker.com/blog/announcing-docker-compose-watch-ga-release/)). I've taken the example of the compose documentation page : https://docs.docker.com/compose/file-watch/ 

> Also why cargo install --path . since it's a ""devcontainer"" ?

Even in a development container, we need to install the cargo dependencies. Running `cargo install` results in an error that can be resolved by using `cargo install --path .`
However, as I'm not a Rust expert, there might be a more optimal approach.

lilymonad replied:
Fair enough, I've got poor windows dev experience I did not see the WSL problem. Thanks for documenting it !

PS: Man it's insane, an entire layer rebuild is more performant than binding files through a vm ? ...

lilymonad replied:
> Even in a development container, we need to install the cargo dependencies. 

Well `cargo install` is for installing the built executable somewhere, it is just doing `cargo build` + copying artefacts.

Same as `cargo run`, it will `cargo build` before running the executable if it was not previously built.
A simple `cargo build` should suffice, and I'm not even sure why you need to build it before `cargo watch`

pbrissaud replied:
Maybe you're right on this ! Will test on my side

pbrissaud committed:
chore: configure port binding with env

Add 2 new env for compose API_PORT and WEB_PORT to define the local port of these services

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud commented:
Added  2 new env for compose `API_PORT` and `WEB_PORT` to define the local port of these services

lilymonad commented:
Ach too bad I dumped docker for podman, I can't test the feature (no --watch for podman-compose) ... well I guess a nix devShell wouldn't be a bad thing then lol silly me :p

pbrissaud commented:
Both solutions can very well coexist to include as many people as possible!

vasilvestre commented:
Why would you use a Dockerfile.dev ?
Multistage is made to avoid having multiple files and would be great to use here

pbrissaud commented:
Using a multistage build is a viable approach, but having a separate Dockerfile.dev significantly enhances the readability and maintainability of our Docker setup. The distinct files allow developers to differentiate between development and production environments, making it easier to understand and modify the build process for each use case.

pbrissaud committed:
fix: default web port

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud committed:
fix: remove useless instructions on Dockerfile.dev for API

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud committed:
fix: remove useless `npm install` command on Dockerfile.dev

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud committed:
chore: retrigger build on change on Dockerfile.dev

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud commented:
Some fixes :
- Fix the default port for the web app
- Remove useless instructions on Dockerfile.dev (ty @lilymonad for the Rust side)
- Retrigger build when dockerfile changes

vasilvestre commented on web/Dockerfile.dev:
```diff
@@ -0,0 +1,13 @@
+
+COPY package*.json .
+
+RUN npm ci
```
You can add --audit false and --fund false for extra performance and readability

pbrissaud committed:
Add option to npm install for extra performance and readability

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

ferdodo commented on compose.yaml:
```diff
@@ -6,23 +6,13 @@ services:
     build: ./web
     ports: 
-      - 3000:3000
+     - ${WEB_PORT:-8080}::3000
```
there's a double colon here, I think you need only one

pbrissaud replied:
Thanks you !

pbrissaud committed:
fix: typo

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud committed:
fix: delete override part

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

Lauwed committed:
fix: Adding draft README doc

ferdodo commented on web/Dockerfile.dev:
```diff
@@ -0,0 +1,13 @@
+FROM node:18-alpine
```
Can you also define a environment variable for node version ?

this could be useful as mentioned here https://github.com/DevGirl-Team/tamagokill/pull/7#discussion_r1629225992

pbrissaud replied:
Of course! In dev and prod images?

vasilvestre replied:
You should run the same version for dev and prod. Take a look at https://github.com/DevGirl-Team/tamagokill/blob/2-ajout-environnents-dev-et-prod-au-docker/api/Dockerfile#L2
You can use the version passed as ARG or define it yourself for further testing

pbrissaud replied:
I just set up a build arg `NODE_VERSION` in both dockerfiles to change the node version used in the container. 

The default version is 18. 

You can change it in compose file like that : 

```
build:
  ...
  args:
    NODE_VERSION: 20
```

and the value can be an environment variable.

pbrissaud replied:
You can set a NODE_VERSION environment variable to change the node version used to build the image. 

```
web:
  build:
    context: ./web
    dockerfile: Dockerfile.dev
    args:
      - NODE_VERSION=${NODE_VERSION:-18}
```

The default is `18`. You can use any value that matches a node docker tag with a `-alpine` suffix: https://hub.docker.com/_/node/tags?page=&page_size=&ordering=&name=alpine

pbrissaud committed:
chore: Update Node.js version in Dockerfile to use the specified ARG

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

pbrissaud committed:
feat: Customize Node.js version in compose files

Signed-off-by: Paul Brissaud <pbrissaud.pro@gmail.com>

Lauwed committed:
fix: Change the WEB_PORT in compose.yaml
